File: 5. Longest Palindromic Substring/solution.cpp
/*
    Runtime: 16 ms, faster than 83.54% of C++ online submissions for Longest Palindromic Substring.
    Memory Usage: 8.7 MB, less than 90.14% of C++ online submissions for Longest Palindromic Substring.
*/

class Solution {
public:
    void ToCount(const string& s, int& max_size, int& pos, int num, int i){
        int j = 1, size = 2 + num;
        while(j <= i && j < s.size() - i && s[i - j] == s[i + 1 + j + num]){
            size += 2;
            j++;
        }
        if (size > max_size){
            max_size = size;
            pos = i - j + 1;
        }
    }
    
    string longestPalindrome(string s) {
        if (s.empty())
            return "";
        int max_size = 1, pos = 0;
        for(int i = 0; i + 1 < s.size(); ++i){
            if (s[i] == s[i+1])
                ToCount(s, max_size, pos, 0, i);
            if (i + 2 < s.size())
                if (s[i] == s[i + 2])
                    ToCount(s, max_size, pos, 1, i);
        }
        return s.substr(pos, max_size);
    }
};

File: 5. Longest Palindromic Substring/solution.py
# Runtime: 904 ms, faster than 76.37% of Python3 online submissions for Longest Palindromic Substring.
# Memory Usage: 13.1 MB, less than 76.49% of Python3 online submissions for Longest Palindromic Substring.

class Solution:
    def longestPalindrome(self, s: str) -> str:
        res = ""
        for i in range(len(s)):
            tmp = self.to_count(s, i, i)
            res = tmp if len(tmp) > len(res) else res
            tmp = self.to_count(s, i, i+1)
            res = tmp if len(tmp) > len(res) else res
        return res
    
    @staticmethod
    def to_count(s, lhs, rhs):
        while lhs >= 0 and rhs < len(s) and s[lhs] == s[rhs]:
            lhs -= 1; rhs += 1
        return s[lhs+1:rhs]
    
